src/abstractdataframe/show.jl

```diff
@@ -573,61 +573,104 @@ function _show(io::IO,
+        if truncate == 0
+            maximum_columns_width[i] = 0
+        else
+            maximum_columns_width[i] =
+                max(truncate + 1, names_len[i], textwidth(types[i]))
         end
     end
-    dsize = displaysize(io)
-    availableheight = dsize[1] - 7
-    nrowssubset = fld(availableheight, 2)
-    bound = min(nrowssubset - 1, nrows)
-    if allrows || nrows <= availableheight
-        rowindices1 = 1:nrows
-        rowindices2 = 1:0
+
+    # Check if the user wants to display a summary about the DataFrame that is
+    # being printed. This will be shown using the `title` option of
+    # `pretty_table`.
+    title = summary ? Base.summary(df) : ""
+
+    # If `rowid` is not `nothing`, then we are printing a data row. In this
+    # case, we will add this information using the row name column of
+    # PrettyTables.jl. Otherwise, we can just use the row number column.
+    if (rowid === nothing) || (size(df, 2) == 0)
+        show_row_number = true
+        row_names = nothing
     else
-        rowindices1 = 1:bound
-        rowindices2 = max(bound + 1, nrows - nrowssubset + 1):nrows
+        size(df, 1) != 1 &&
+            throw(ArgumentError("rowid may be passed only with a single row data frame"))
+        show_row_number = false
+        row_names = [string(rowid)]
     end
-    maxwidths = getmaxwidths(df, io, rowindices1, rowindices2, rowlabel, rowid,
-                             eltypes, buffer, truncstring)
-    width = getprintedwidth(maxwidths)
-    showrows(io, df, rowindices1, rowindices2, maxwidths, splitcols, allcols,
-             rowlabel, summary, eltypes, rowid, buffer, truncstring)
-    return
+
+    # Print the table with the selected options.
+    pretty_table(io, df, vcat(names,types);
+                 alignment                   = :l,
+                 compact_printing            = get(io, :compact, true),
+                 continuation_row_alignment  = :l,
+                 crop                        = crop,
+                 crop_num_lines_at_beginning = 2,
+                 formatters                  = (_pretty_tables_formatter,),
+                 highlighters                = (_PRETTY_TABLES_HIGHLIGHTER,),
+                 maximum_columns_width       = maximum_columns_width,
+                 newline_at_end              = false,
+                 nosubheader                 = !eltypes,
+                 row_name_alignment          = :l,
+                 row_name_crayon             = Crayon(),
+                 row_name_column_title       = string(rowlabel),
+                 row_names                   = row_names,
+                 row_number_alignment        = :l,
+                 row_number_column_title     = string(rowlabel),
+                 show_row_number             = show_row_number,
+                 tf                          = dataframe,
+                 title                       = title,
+                 vlines                      = [1],
```

Comment:
No, it's not! All the options are those that are not default in PrettyTables or there are needed due to the configuration of the output (like highlights).

Reply:
OK. Though `dataframe` isn't defined anywhere?

Reply:
`dataframe` is a table type of PrettyTables to match the design of DataFrames before this PR.

Reply:
Got it. It's a bit weird though for `dataframe` to differ from the default printing of `DataFrame`.

Reply:
I have now checked what you export in PrettyTables.jl. The list is long and some of the names are potentially common.

Out of curiosity - have you received complaints on such things from end users? This is in general a PrettyTables.jl decision of course, but e.g. @quinnj in Tables.jl decided not to export common names, but require them to be qualified with `Table.` prefix.

Reply:
Another common strategy is to accept `Union{Symbol, TypeYouExpect}` and just use `Symbol` for predefined operations. Usually `TypeYouEpect` is a function, so there is no conflict with `Symbol`. In this way you can avoid exporting things.

Reply:
> Got it. It's a bit weird though for `dataframe` to differ from the default printing of `DataFrame`.

This type only defines the border structure and if the horizontal or vertical lines will be printed. It is almost a copy of the default type in PrettyTables. I decided to make a separate type to DataFrames so that a change in the default printing design of PrettyTables will not break DataFrames.

> Out of curiosity - have you received complaints on such things from end users? This is in general a PrettyTables.jl decision of course, but e.g. @quinnj in Tables.jl decided not to export common names, but require them to be qualified with `Table.` prefix.

Actually I have never though about that. Also nobody complained about a possible crash. I could verify this in the future, but removing those export for now without a deprecation period will lead to massive breakage of existing code (which I constantly use in my work).

Reply:
> removing those export for now without a deprecation period will lead to massive breakage of existing code (which I constantly use in my work

sure. However, a typical deprecation path would be:

Step 1: allow both `dataframe` and a `Symbol` `:dataframe` in the deprecation period

Step 2: after deprecation period unexport `dataframe` and either:

Step 3a: replace all `dataframe` occurences by prepending them with `:` (this is a one-line regex replace), or:

Step 3b: add `import PrettyTables: dataframe` statement at the top of the script

(I do not say you have to do it now - we do not export PrettyTables.jl - I am just saying that here handling the deprecation is relatively easy if you wend this way)

Reply:
@bkamins by the way, I have never released a version with `dataframes`. I created just to protect DataFrames from changes in the default table design of PrettyTables. If you like, I can define this constant inside DataFrames and remove this variable from PrettyTables. Should I do this?

Reply:
As you prefer. I only wanted to hint you that when PrettyTables.jl goes mainstream then being careful what is exported will become a critical issue. In DataFrames.jl the naming is one of the things that we often spend more time than implementation as users have very complex needs here.

I assume that you know the design of PrettyTables.jl best, so you can make a sound judgement what is the best approach given the comments I made (as they are from experience of maintaining DataFrames.jl).

Reply:
OK I hadn't realized the point of this `dataframe` type was to protect DataFrames. In that case, why not move more properties to that type so that other users can get a printing which is close enough what we do here?

At any rate I agree with @bkamins that it's often better to use symbols. If `dataframe` hasn't been included in a release yet I'd recommend not exporting it (I'm a bit worried about the confusion between `dataframe` and `DataFrame`).